### PR TITLE
Establish testframework for automatic testing of authentication and authorization

### PR DESCRIPTION
**Pull request by jennyhougen**
Features:
- Setup fake authentication schemes. One simulating cookie (need to go through if we want this) and one setting fake user based on a TTT generated token
- Create authentication and authorization tests
- Adding testguideline from what have been discussed. Need to decide on where this should be and also completed.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -10,14 +10,6 @@ indent_size = 4
 dotnet_diagnostic.CS1591.severity = none
 dotnet_sort_system_directives_first = true
```
Comment (jennyhougen) [inline .editorconfig:12]: Will be gone after stylecop branch to Geir Marius. Needed to do this to get my intellisence to work..

File: Fhi.HelseId.Integration.Tests/HelseId.Web/Tests/AuthenticationTests.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                services.AddHelseIdWebAuthentication(config)
+               .Build();
+            });
+
+
+            var client = appFactory.CreateClient();
+            var response = await client.GetAsync("/api/test");
+
+            Assert.That(response.StatusCode, Is.EqualTo(HttpStatusCode.Unauthorized));
+            var queryParams = HttpUtility.ParseQueryString(response.Headers.Location is not null ? response.Headers.Location.Query : string.Empty );
```
Comment (GeirMG) [inline Fhi.HelseId.Integration.Tests/HelseId.Web/Tests/AuthenticationTests.cs:31]: Avoid logic in tests, if response.Headers.Location is null it should crash/fail ?

File: Fhi.HelseId.Integration.Tests/TestFramework/AuthenticationSchemes/TestAuthenticationScheme/TestAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                return Task.FromResult(AuthenticateResult.Success(ticket));
+            }
+
+            if (Options.ClaimsIssuer is not null) 
+            {
+                var ticket = new AuthenticationTicket(new ClaimsPrincipal(new ClaimsIdentity(Options.UserClaims, "TestAuthentication")), Scheme.Name);
+                return Task.FromResult(AuthenticateResult.Success(ticket));
+            }
+
+            return Task.FromResult(AuthenticateResult.Fail(new Exception("")));
```
Comment (GeirMG) [inline Fhi.HelseId.Integration.Tests/TestFramework/AuthenticationSchemes/TestAuthenticationScheme/TestAuthenticationHandler.cs:41]: Don't fail with genric Exception?

Ref. General guideline / Treat tests as production code

File: Fhi.HelseId.Integration.Tests/TestFramework/NHNTTT/Dtos/DPoPProofParameters.cs [hunk 4]
```diff
@@ -8,7 +8,7 @@ internal record DPoPProofParameters
 
         public string? PrivateKeyForProofCreation { get; set; }
 
-        public InvalidDPoPProofParameters? InvalidDPoPProofParameters { get; set; }
+        //public InvalidDPoPProofParameters? InvalidDPoPProofParameters { get; set; }
```
Comment (GeirMG) [inline Fhi.HelseId.Integration.Tests/TestFramework/NHNTTT/Dtos/DPoPProofParameters.cs:11]: Commented out code - remove?

File: Fhi.HelseId.Integration.Tests/TestFramework/Extensions/IConfigurationRootExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                configuration.Authority,
+                configuration.ClientId,
+                JsonSerializer.Serialize(configuration.Scopes));
+        }
+
+        private static IConfigurationRoot CreateHelseIdWebConfiguration(string securityLevel = "3", bool authUse = true, string authority = "https://helseid-sts.test.nhn.no/", string clientId = "1224", string scopes = "fhi:api-scope")
+        {
+            var appsettingsConfig = new Dictionary<string, string?>
+                {
+                    { "HelseIdWebKonfigurasjon:AuthUse", authUse.ToString() },
```
Comment (GeirMG) [inline Fhi.HelseId.Integration.Tests/TestFramework/Extensions/IConfigurationRootExtensions.cs:30]: Some magic strings, I see that
$"{nameof(HelseIdWebKonfigurasjon)}:{nameof(HelseIdWebKonfigurasjon.AuthUse)}"
is used elsewhere - good idea?

File: Fhi.HelseId.Integration.Tests/TestFramework/Readme.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        var appFactory = new TestWebApplicationFactory(testConfiguration, services =>
+        {
+            ..   
+        });
+
+        // Act
+        var client = appFactory.CreateClient();
+        var response = await client.GetAsync("api/person");
+
+        //Assert
```
Comment (GeirMG) [inline Fhi.HelseId.Integration.Tests/TestFramework/Readme.md:30]: Nit: // Assert (will be caught by StyleCop when implemented)

Several occurrences in this file.

File: Fhi.HelseId.Integration.Tests/TestFramework/Readme.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+
+### Using test framework
+
+The Testframework uses [MS integrationtest framework](https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-9.0) to interfare with the request response pipeline. 
+This allows to test different configurations and to simulate different usages by mocking out parts of the services. 
+
+```
+ [Test]
+        public async Task SampleSetup_WithOverideOpenIdConnectConfigurationInTests()
```
Comment (GeirMG) [inline Fhi.HelseId.Integration.Tests/TestFramework/Readme.md:47]: Missing the then-part?

File: Fhi.HelseId.Integration.Tests/TestFramework/Readme.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+```
+
+
+#### Sample of fake authentication using cookie
+
+
+```
+[Test]
+public async Task Overriding_CookieAuth()
```
Comment (GeirMG) [inline Fhi.HelseId.Integration.Tests/TestFramework/Readme.md:138]: Overriding_CookieAuth - and then what? :)

**Review by GeirMG — COMMENTED** [review]
There are some changes in the implementation at the very end of this PR. Just a heads up since the PR gives the impression that this will establish integration tests.

File: Fhi.HelseId.Integration.Tests/HelseId.Api/Tests/DPoPTests.cs [hunk 9]
```diff
@@ -17,11 +16,8 @@ public class DPoPTests
     [Test]
     public async Task ApiCallWithDpopToken_ApiAcceptsBothDpopAndBearer_Returns200Ok()
     {
-        var config = HelseIdApiKonfigurasjonExtensions.CreateHelseIdApiKonfigurasjon(
-            allowDPoPTokens: true, 
-            requireDPoPTokens: false,
-            audience: "fhi:helseid.testing.api",
-            allowedScopes: "fhi:helseid.testing.api/all");
+        var config = HelseIdApiKonfigurasjonBuilder.Create.DefaultValues(audience: "fhi:helseid.testing.api", allowedScopes: "fhi:helseid.testing.api/all")
```
[the remark this answers is not in the document]
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/HelseId.Api/Tests/DPoPTests.cs:19]: Agree

File: Fhi.HelseId.Integration.Tests/HelseId.Api/Setup/HelseIdKonfigurasjonTestDefaults.cs [hunk 10]
```diff
@@ -3,27 +3,39 @@
 
 namespace Fhi.HelseId.Integration.Tests.HelseId.Api.Setup
 {
-    public static class HelseIdApiKonfigurasjonExtensions
+    internal class HelseIdApiKonfigurasjonBuilder
```
[the remark this answers is not in the document]
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/HelseId.Api/Setup/HelseIdKonfigurasjonTestDefaults.cs:6]: Will be public eventually.

File: Fhi.HelseId.Integration.Tests/HelseId.Web/Setup/HelseIdWebKonfigurasjonExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
+﻿
+namespace Fhi.HelseId.Web
+{
+    internal class HelseIdWebKonfigurasjonBuilder
```
[the remark this answers is not in the document]
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/HelseId.Web/Setup/HelseIdWebKonfigurasjonExtensions.cs:4]: Same as above

File: Fhi.HelseId.Integration.Tests/TestFramework/AuthenticationSchemes/TestAuthenticationScheme/TestAuthenticationHandler.cs (hunk 3 above)
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/TestFramework/AuthenticationSchemes/TestAuthenticationScheme/TestAuthenticationHandler.cs:41]: Will fix

File: Fhi.HelseId.Integration.Tests/HelseId.Web/Tests/AuthenticationTests.cs (hunk 2 above)
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/HelseId.Web/Tests/AuthenticationTests.cs:31]: Will fix

File: Fhi.HelseId.Integration.Tests/TestFramework/NHNTTT/Dtos/DPoPProofParameters.cs (hunk 4 above)
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/TestFramework/NHNTTT/Dtos/DPoPProofParameters.cs:11]: Add //// use when intention

File: Fhi.HelseId.Integration.Tests/TestFramework/Extensions/IConfigurationRootExtensions.cs (hunk 5 above)
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/TestFramework/Extensions/IConfigurationRootExtensions.cs:30]: Agree

File: Fhi.HelseId.Integration.Tests/TestFramework/Readme.md (hunk 6 above)
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/TestFramework/Readme.md:30]: Illustrating AAA pattern in readme guideline. Closing this.

File: Fhi.HelseId.Integration.Tests/TestFramework/AuthenticationSchemes/CookieAuthenticationScheme/FakeCookieManager.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
+﻿using Fhi.TestFramework.NHNTTT;
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Http;
+
+namespace Fhi.TestFramework.AuthenticationSchemes.CookieScheme
```
[the remark this answers is not in the document]
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/TestFramework/AuthenticationSchemes/CookieAuthenticationScheme/FakeCookieManager.cs:5]: Intended. Will go through in the meeting.

File: Fhi.HelseId.Integration.Tests/HelseId.Web/Tests/AuthenticationTests.cs [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+            Assert.That(queryParams["redirect_uri"], Is.EqualTo("http://localhost/signin-callback"));
+        }
+
+        /// <summary>
+        /// Test signin callback with token
+        /// https://github.com/dotnet/aspnetcore/blob/81a2bab8704d87d324039b42eb1bab0d977f25b8/src/Security/Authentication/test/OpenIdConnect/OpenIdConnectEventTests_Handler.cs
+        /// </summary>
+        /// <returns></returns>
+        [Test]
+        [Ignore("Will be implemented later")]
```
[the remark this answers is not in the document]
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/HelseId.Web/Tests/AuthenticationTests.cs:43]: Same as above

File: Fhi.HelseId.Integration.Tests/TestFramework/Readme.md [hunk 14]
```diff
@@ -0,0 +1,197 @@
+﻿# Testing guideline
```
Comment (jennyhougen) [inline Fhi.HelseId.Integration.Tests/TestFramework/Readme.md:1]: Move to wiki

File: Fhi.HelseId.Integration.Tests/TestFramework/AuthenticationSchemes/CookieAuthenticationScheme/FakeCookieManager.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/TestFramework/AuthenticationSchemes/CookieAuthenticationScheme/FakeCookieManager.cs:5]: As agrred on in the meeting that we will test this approach

File: Fhi.HelseId.Integration.Tests/HelseId.Web/Tests/AuthenticationTests.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/HelseId.Web/Tests/AuthenticationTests.cs:43]: Created a case

File: Fhi.HelseId.Integration.Tests/TestFramework/NHNTTT/Dtos/DPoPProofParameters.cs [hunk 15]
```diff
@@ -8,6 +8,8 @@ internal record DPoPProofParameters
 
         public string? PrivateKeyForProofCreation { get; set; }
 
-        public InvalidDPoPProofParameters? InvalidDPoPProofParameters { get; set; }
+        //// Removed the implementation as it was wrong. This is a property in TTT that
```
Comment (GeirMG) [inline Fhi.HelseId.Integration.Tests/TestFramework/NHNTTT/Dtos/DPoPProofParameters.cs:11]: Nit: Normal comments can have //
Commented out code should have //// if it was intentional / has meaning

**Review by GeirMG — APPROVED** [review]
This is a good start, and I look forward to where this takes us.

This PR feels "sharper" somehow. Tighter, cleaner. I like it.

File: Fhi.HelseId.Integration.Tests/TestFramework/NHNTTT/Dtos/DPoPProofParameters.cs (hunk 15 above)
Comment (jennyhougen) [reply Fhi.HelseId.Integration.Tests/TestFramework/NHNTTT/Dtos/DPoPProofParameters.cs:11]: Fikser det i neste.

**Pull request merged**